buttplug/src/server/device/protocol/mizzzee_v2_1.rs

```diff
@@ -0,0 +1,156 @@
+      ProtocolHandler,
+      ProtocolIdentifier,
+      ProtocolInitializer,
+    },
+    ServerDeviceIdentifier,
+  },
+  util::{async_manager, sleep},
+};
+use async_trait::async_trait;
+use std::sync::atomic::{AtomicU32, AtomicU8, Ordering};
+use std::{sync::Arc, time::Duration};
+
+generic_protocol_initializer_setup!(MizzZeeV2_1, "mizzzee-v2.1");
+
+#[derive(Default)]
+pub struct MizzZeeV2_1Initializer {}
+
+#[async_trait]
+impl ProtocolInitializer for MizzZeeV2_1Initializer {
+  async fn initialize(
+    &mut self,
+    hardware: Arc<Hardware>,
+    _: &ProtocolDeviceAttributes,
+  ) -> Result<Arc<dyn ProtocolHandler>, ButtplugDeviceError> {
+    Ok(Arc::new(MizzZeeV2_1::new(hardware)))
+  }
+}
+
+// Time between MizzZee2 update commands, in milliseconds.
+const MIZZZEE2_COMMAND_DELAY_MS: u64 = 20;
+
+// Time between MizzZee2 keep vibrating commands, in milliseconds.
+const MIZZZEE2_COMMANDS_KEEP_VIBRATING: u64 = 200;
+
+// Amount of commands that can be skipped without stopping the device.
+const MIZZZEE2_COMMANDS_TO_SKIP: u8 = (MIZZZEE2_COMMANDS_KEEP_VIBRATING / MIZZZEE2_COMMAND_DELAY_MS - 1) as u8;
+
+fn handle_scale(scale: f32) -> f32 {
+  if scale == 0.0 { return 0.0; }
+  scale * 0.7 + 0.3
+}
+
+fn scalar_to_vector(scalar: u32) -> Vec<u8> {
+  const HEADER: [u8; 3] = [0x03, 0x12, 0xf3];
+  const FILL_VEC: [u8; 6] = [0x00, 0xfc, 0x00, 0xfe, 0x40, 0x01];
+
+  let scale: f32 = handle_scale(scalar as f32 / 1000.0) * 1023.0;
+  let modded_scale: u16 = ((scale as u16) << 6) | 60;
+  
+  let bytes = modded_scale.swap_bytes().to_be_bytes();
+
+  let mut data: Vec<u8> = Vec::new();
+  data.extend_from_slice(&HEADER);
+  data.extend_from_slice(&FILL_VEC);
+  data.extend_from_slice(&bytes);
+  data.extend_from_slice(&FILL_VEC);
+  data.extend_from_slice(&bytes);
+  data.push(0x00);
+
+  data
+}
+
+async fn vibration_update_handler(
+  device: Arc<Hardware>,
+  loops_skipped_holder: Arc<AtomicU8>,
+  last_scalar_holder: Arc<AtomicU32>,
+  current_scalar_holder: Arc<AtomicU32>,
+) {
+  info!("Entering MizzZee2 Control Loop");
+  loop {
+    sleep(Duration::from_millis(MIZZZEE2_COMMAND_DELAY_MS)).await;
+
+    let loops_skipped = loops_skipped_holder.load(Ordering::Relaxed);
+    let last_scalar = last_scalar_holder.load(Ordering::Relaxed);
+    let current_scalar = current_scalar_holder.load(Ordering::Relaxed);
+
+    if last_scalar == current_scalar && loops_skipped < MIZZZEE2_COMMANDS_TO_SKIP {
+      loops_skipped_holder.store(loops_skipped + 1, Ordering::Relaxed);
+      continue;
+    }
+
+    loops_skipped_holder.store(0, Ordering::Relaxed);
+    last_scalar_holder.store(current_scalar, Ordering::Relaxed);
+
+    if device
+      .write_value(&HardwareWriteCmd::new(
+        Endpoint::Tx,
+        scalar_to_vector(current_scalar),
+        true
+      ))
+      .await
+      .is_err() { break; }
+
+    info!("MZ2 scalar: {}", current_scalar);
+  }
+  info!("MizzZee2 control loop exiting, most likely due to device disconnection.");
```

Comment:
I know these are `info!` in the MysterVibe protocol, but as you've already commented, they're loud. Maybe `trace!`?

Reply:
I thik you're right about logs inside of the loop, but i don't think that logs outside of it are redundant